### ISSUE TITLE
Sidebar Display Time Has Slowed During Moonshine Start

### ISSUE THREAD

**Issue by rat-moonshine** [Moonshine-IDE/Moonshine-IDE#1002]
From @JustinProminic 

---------------------------------

The launch time for Moonshine to show the sidebar after the Home screen has displayed has increased drastically in the past few nights of development builds.

I suspect this has to do with the External Editors or something.

I do not have big project files loaded and don't think it was the launch of the Language Servers that was slowing it down in this case.

Please keep an eye on exactly what tasks we are starting before the UI is responsive to the user after launch.

I think we are going to need a GitHub issue or something just to track the absolute must checks during startup.  If the issue is figuring out whether to display the Getting Started page, that should happen in a separate thread of execution so we do not block the main UI thread.

---------------------------------

**Comment by rat-moonshine** [comment]
**Ongoing test-stats**

Startup test with projects opened:
GBAuthDevelopment, MoonshineDesktop, MoonshineSharedCore, ProminicNativeWeb (Royale)

Scenario | Second(s) to show sidebar with projects tree
------------|-----------------------
LS support | 8
LS disabled | 7
LS disabled, `FSWatcherPlugin` disabled | 7

**Comment by rat-moonshine** [comment]
Based on the above test-findings, 10/11 seconds may not be that bad for a 6 large size of projects on an average of max.2 seconds to each. Disabling LS also didn't shown any changes to waiting seconds, so LS not having any suspicion to the delay, at this moment.

However, we all agreed that an asynchronous UI would be more expected here than freezing the interface for 10+ seconds. We should work on that.

**Comment by JoelProminic** [comment]
What is running when projects open.  Here is a quick list we discussed during the meeting:
- language server
- file list - I think we are only loading the first level of files currently
- reopen tree state
- file watcher
- reopen files

**Comment by rat-moonshine** [comment]
**Note: This update is not to find reason of delays but serialising the project opening processes**

Beside that, I think the earlier [findings](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1002#issuecomment-1082907406) on project start time are reasonable.

I did couple of changes, introducing some new design/interfaces etc. to serialising the execution processes resides in different plugin-classes, centrally, during one or multiple project start. 

Following are the plugin/classes responsible to execute their respective processes during a project start/import:
- ProjectPlugin
- MenuPlugin
- WorkspacePlugin
- RecentlyOpenedPlugin
- SVNPlugin
- ILanguageServerBridgeImp
- FSWatcherPlugin

These all processes now managed by [ProjectStarter](https://github.com/Moonshine-IDE/Moonshine-IDE/blob/features/issue_1002_async_sidebar/ide/MoonshineSharedCore/src/actionScripts/plugin/project/ProjectStarter.as) in orderly way, and subscribed by the above plugin/classes.

Now one can see multiple projects being added into sidebar one by one, after having their respective opening processes being complete. 

Additionally, you can also see a new progress-output at bottom-right status-bar giving information on queued projects and displayed until all the process completion. (the queued number/label has some problem I need to adjust)

![image](https://user-images.githubusercontent.com/24370870/162417287-978c3386-e2b2-40d9-9dc0-aba0829a2ed3.png)

**Comment by JoelProminic** [comment]
To test this, I tried opening a workspace with the Moonshine source projects:

![image](https://user-images.githubusercontent.com/24532731/162501477-4b340811-b5a1-4c62-885d-a1ac735e424b.png)

I timed from when I opened the workspace (or canceled the update prompt after restarting Moonshine) until the projects seemed to all be opened and the busy wheel cleared.   I did two tests for each build.  The times are not very precise, but they should give an idea of 

Build | Test 1 | Test 2
------|--------|-------
master | 35 | 40
1002 branch, language servers started in sequence | 45 | 60
1002 branch, language servers started at the end | 25 | 40

Personally, I didn't like having the projects load one\-by\-one like this.  It felt confusing and distracting.  This might work better if we had placeholders for the projects in the sidebar with busy icons.  I see the progress in the status bar, but I missed this in my first couple tests.

If we do keep this solution, we need to be careful about reopened files stealing focus as the projects opened.  Even if project and file I wanted to work on opened first, I wouldn't be able to start working on it because Moonshine would constantly switch to other tabs.

**Comment by JoelProminic** [comment]
Unfortunately, I see that the Bamboo build is failing:

    11-Apr-2022 10:27:45 	    [mxmlc] .../MS/g-BBMGO-JOB1/Moonshine-IDE/ide/MoonshineSharedCore/src/visualEditor/plugin/ExportDominoToRoyalePlugin.as(96): col: 71 Error: Call to a possibly undefined method fromXMLOnly through a reference with static type converter:DominoConverter.
    11-Apr-2022 10:27:45 	    [mxmlc] 
    11-Apr-2022 10:27:45 	    [mxmlc]                     var surfaceMockup:SurfaceMockup = dominoConverter.fromXMLOnly(dominoXML);
    11-Apr-2022 10:27:45 	    [mxmlc]                                                                       ^
    11-Apr-2022 10:27:45 	    [mxmlc] 
    11-Apr-2022 10:27:45 	    [mxmlc] .../MS/g-BBMGO-JOB1/Moonshine-IDE/ide/MoonshineDESKTOPevolved/src/actionScripts/plugins/as3project/CreateProject.as(1003): col: 10 Warning: Duplicate variable definition.
    11-Apr-2022 10:27:45 	    [mxmlc] 
    11-Apr-2022 10:27:45 	    [mxmlc]                                         var newRoyaleMxml:FileLocation =  targetFolder.resolvePath("src"+File.separator+pvo.projectName + ".mxml");
    11-Apr-2022 10:27:45 	    [mxmlc]                                             ^
    11-Apr-2022 10:27:45 	    [mxmlc] 
    11-Apr-2022 10:27:45 	    [mxmlc] .../MS/g-BBMGO-JOB1/Moonshine-IDE/ide/MoonshineDESKTOPevolved/src/actionScripts/plugins/as3project/CreateProject.as(1015): col: 10 Warning: Duplicate variable definition.
    11-Apr-2022 10:27:45 	    [mxmlc] 
    11-Apr-2022 10:27:45 	    [mxmlc]                                         var newRoyaleTemplate:FileLocation =  targetFolder.resolvePath(pvo.projectName + ".as3proj");

I think the branch needs a merge from master for some recent Domino Visual Editor updates.

**Comment by JoelProminic** [comment]
I merged master into the branch myself to keep the tests moving. Make sure to pull the branch before you make updates.

The merge resolved the Bamboo error.  If someone else wants to test the updated builds for this issue, the branch build is available here:  https://moonshine-ide.com/downloads/releases/dev/macos/branches/features/issue_1002_async_sidebar

However, when I switched the workspace, I got these errors:

    : Error #1009
    : TypeError: Error #1009
    : 	at actionScripts.plugin.workspace::WorkspacePlugin/onProjectAdded()
    : 	at flash.events::EventDispatcher/dispatchEvent()
    : 	at actionScripts.plugin.project::ProjectPlugin/handleRemoveProject()
    : 	at flash.events::EventDispatcher/dispatchEvent()
    : 	at Function/<anonymous>()
    : 	at Function/actionScripts.utils:UtilsCore/closeAllRelativeEditors/actionScripts.utils:onModListClosed()
    : 	at actionScripts.utils::UtilsCore$/closeAllRelativeEditors()
    : 	at components.views.project::TreeView/onFileDeletedOnServer()
    : 	at components.views.project::TreeView/onCloseProjectRequest()
    : 	at flash.events::EventDispatcher/dispatchEvent()
    : 	at actionScripts.plugin.workspace::WorkspacePlugin/closeAllEditorAsync()
    : 	at actionScripts.plugin.workspace::WorkspacePlugin/handleRemoveProjectByPath()
    : 	at actionScripts.plugin.workspace::WorkspacePlugin/closeAllEditorAsync()
    : 	at actionScripts.plugin.workspace::WorkspacePlugin/handleLoadWorkspaceEvent()
    : 	at flash.events::EventDispatcher/dispatchEvent()
    : 	at feathers.controls::ScrollContainer/dispatchEvent()
    : 	at moonshine.plugin.workspace.view::LoadWorkspaceView/loadWorkspaceButton_triggerHandler()
    : 	at flash.events::EventDispatcher/dispatchEvent()
    : 	at feathers.events::TriggerEvent$/dispatchFromMouseEvent()
    : 	at feathers.utils::PointerTrigger/pointerTrigger_target_clickHandler()
    :
    : Click here to Report a Bug
    : Error #1009
    : TypeError: Error #1009
    : 	at actionScripts.plugin.recentlyOpened::RecentlyOpenedPlugin/onProjectAdded()
    : 	at flash.events::EventDispatcher/dispatchEvent()
    : 	at actionScripts.plugin.project::ProjectPlugin/handleRemoveProject()
    : 	at flash.events::EventDispatcher/dispatchEvent()
    : 	at Function/<anonymous>()
    : 	at Function/actionScripts.utils:UtilsCore/closeAllRelativeEditors/actionScripts.utils:onModListClosed()
    : 	at actionScripts.utils::UtilsCore$/closeAllRelativeEditors()
    : 	at components.views.project::TreeView/onFileDeletedOnServer()
    : 	at components.views.project::TreeView/onCloseProjectRequest()
    : 	at flash.events::EventDispatcher/dispatchEvent()
    : 	at actionScripts.plugin.workspace::WorkspacePlugin/closeAllEditorAsync()
    : 	at actionScripts.plugin.workspace::WorkspacePlugin/handleRemoveProjectByPath()
    : 	at actionScripts.plugin.workspace::WorkspacePlugin/closeAllEditorAsync()
    : 	at actionScripts.plugin.workspace::WorkspacePlugin/handleLoadWorkspaceEvent()
    : 	at flash.events::EventDispatcher/dispatchEvent()
    : 	at feathers.controls::ScrollContainer/dispatchEvent()
    : 	at moonshine.plugin.workspace.view::LoadWorkspaceView/loadWorkspaceButton_triggerHandler()
    : 	at flash.events::EventDispatcher/dispatchEvent()
    : 	at feathers.events::TriggerEvent$/dispatchFromMouseEvent()
    : 	at feathers.utils::PointerTrigger/pointerTrigger_target_clickHandler()
    :
    : Click here to Report a Bug
    : Error #1009
    : TypeError: Error #1009
    : 	at actionScripts.plugins.fswatcher::FSWatcherPlugin/onProjectAdded()
    : 	at flash.events::EventDispatcher/dispatchEvent()
    : 	at actionScripts.plugin.project::ProjectPlugin/handleRemoveProject()
    : 	at flash.events::EventDispatcher/dispatchEvent()
    : 	at Function/<anonymous>()
    : 	at Function/actionScripts.utils:UtilsCore/closeAllRelativeEditors/actionScripts.utils:onModListClosed()
    : 	at actionScripts.utils::UtilsCore$/closeAllRelativeEditors()
    : 	at components.views.project::TreeView/onFileDeletedOnServer()
    : 	at components.views.project::TreeView/onCloseProjectRequest()
    : 	at flash.events::EventDispatcher/dispatchEvent()
    : 	at actionScripts.plugin.workspace::WorkspacePlugin/closeAllEditorAsync()
    : 	at actionScripts.plugin.workspace::WorkspacePlugin/handleRemoveProjectByPath()
    : 	at actionScripts.plugin.workspace::WorkspacePlugin/closeAllEditorAsync()
    : 	at actionScripts.plugin.workspace::WorkspacePlugin/handleLoadWorkspaceEvent()
    : 	at flash.events::EventDispatcher/dispatchEvent()
    : 	at feathers.controls::ScrollContainer/dispatchEvent()
    : 	at moonshine.plugin.workspace.view::LoadWorkspaceView/loadWorkspaceButton_triggerHandler()
    : 	at flash.events::EventDispatcher/dispatchEvent()
    : 	at feathers.events::TriggerEvent$/dispatchFromMouseEvent()
    : 	at feathers.utils::PointerTrigger/pointerTrigger_target_clickHandler()
    :
    : Click here to Report a Bug
    : Workspace changed to: Moonshine-IDE


The projects opened almost immediately, but I didn't really trust the timing because of all the errors.  I confirmed that I could expand the project tree and use the language server (it started once I opened an MXML file).

**Comment by rat-moonshine** [comment]
This ready, for some more tests.

**Comment by JoelProminic** [comment]
I noticed this error while testing during the meeting today:

1. Install the latest branch build
2. Switch the workspace with File > Workspace > Switch Workspace.  This works properly
3. Open File > Settings
4. Open Default SDK and change the "Default Apache Flex, Apache Royale, or Feathers SDK" setting.  Save the settings
5. Change back to the original workspace.  BUG:  the original projects don't open
6. Change back to the same workspace that you tried in step 2. BUG:  the projects don't open
7. Restart Moonshine.  BUG: the projects don't reopen
8. Change the workspace again. The projects open properly this time.

I couldn't reproduce it with the master build.  I know this code is still changing, but we'll need to make sure this issue is fixed when we are ready to merge.

Otherwise, the updates seem to be working well.  The load time is much faster, and starting the language servers when a file is opened from the project could be a good performance improvement (though it would also add some delay for cases where the language server is slow to start).

**Comment by rat-moonshine** [comment]
The above problem should be rectified now.

**Comment by rat-moonshine** [comment]
@joshtynjala while working on speeding-up the startup-processes, and since we don't have an `Async` API in AS3, I'm currently working on ordering the execution of project opening processes that affects the startup speed. 

I noticed that after LS having started to a project, opening an LS-editor takes a/two second(s) to fully displayed - this experience varies between the time. I assume this because the editor updates with the LS functionalities/validations/updation etc. 

If I open the editors without the LS support, they shown faster opening than when with the LS.

To improve on this part, I've ordered the executions in following order (for a project):
1. Open previously opened files
2. Open the project to sidebar
3. Trigger any process that should start after the project being added into sidebar
4. Start the LS

While having in this order, I see improve in speed of opening bundle files, but the files those been opened in step.1 not showing LS support after step.4. I assume that the editor instances misses importance references to LS that generates after step.2 or step.4.

I was wondering if there is any way we can update the opened editors from step.1, with the LS references from step.4, and having the LS support available? Please, suggest.

**Comment by joshtynjala** [comment]
> I was wondering if there is any way we can update the opened editors from step.1, with/after the LS references from step.4, and having the LS support available? Please, suggest.

@rat-moonshine The language server is notified of opened files by `LanguageServerTextEditor.dispatchDidOpenEvent()`.

This `dispatchDidOpenEvent()` method will return early if certain requirements aren't met, but it should be called again as each requirement becomes ready. Once all requirements are met, the language server is notified.

One of the times that `dispatchDidOpenEvent()` is called is in a listener for `ProjectEvent.LANGUAGE_SERVER_OPENED`, which is probably where it should meet all requirements in your order of executions.

**Comment by rat-moonshine** [comment]
Unfortunately, starting editors before generating its project-object and corresponding LS, Moonshine fails to open the editor instance as `LanguageServerTextEditor` instance. During this time because of unavailability of `project` and non-population of `ILanguageServerBridgeImp.managers` object, Moonshine fails to [get/generate](https://github.com/Moonshine-IDE/Moonshine-IDE/blob/master/ide/MoonshineDESKTOPevolved/src/actionScripts/impls/ILanguageServerBridgeImp.as#L55) an instance of `LanguageServerTextEditor`. Thus, the editors adds as simple `BasicTextEditor` without any LS support.

I do not want to re-invent the wheel in this area but I understand to successfully start the editor as `LanguageServerTextEditor` it needs to be added after `ILanguageServerBridgeImp.managers` populated with LS references.

I changed the ordering of editor opening as last step after the LS initialised against the project(s). This may show some delay of editor opening, but I don't think I can do much better than this in this stage. 

Please, check.

**Comment by JoelProminic** [comment]
In my tests today, it took 50 seconds to open my Moonshine workspace with 10 open editors.  The projects opened in the sidebar within 10 seconds, but there was a long delay before the editors showed up.

I talked with @rat-moonshine about the strategy for starting the language servers asynchronously.  The basic idea is to trigger the language server to start when an editor is opened, and then have it  call a method to add the [language server integration](https://github.com/Moonshine-IDE/Moonshine-IDE/blob/2d398d1c9d045f45e96075dd9544ac30249f1e31/ide/MoonshineSharedCore/src/actionScripts/ui/editor/LanguageServerTextEditor.as#L54-L64) once the language server was fully started


Here is some psuedocode to help clarify the design.
```
function openEditor(project, file) {
    def languageServer = getLanguageServer(project, file)
    
    if (null == languageServer) {
        openBasicEditor(file)
    }
    else if (languageServer.isStarted()) {
        openLanuageServerEditor(file, languageServer)
    }
    else { // not started
        openLanguageServerEditor(file, null)
        languageServer.addEditorToNotify()
    }

}

function getLanguageServer(project, file) {
    def languageServer = getExistingLanguageServer(project, file)

    if (languageServer != null) {
        return languageServer
    }
    else if (hasLanguageServerSupport(project, file)) {
        languageServer = initializeLanguageServer(project, file) // asynchronous, add to manager list
        return languageServer
    }
    else {
        return null // use basic editor
    }
}
```

We could use similar logic to trigger reloads if the language server needs to be restarted (i.e. configuration changes, forced restarts from #994 interface).

**Comment by joshtynjala** [comment]
> In my tests today, it took 50 seconds to open my Moonshine workspace with 10 open editors. The projects opened in the sidebar within 10 seconds, but there was a long delay before the editors showed up.

> have it call a method to add the [language server integration](https://github.com/Moonshine-IDE/Moonshine-IDE/blob/2d398d1c9d045f45e96075dd9544ac30249f1e31/ide/MoonshineSharedCore/src/actionScripts/ui/editor/LanguageServerTextEditor.as#L54-L64) once the language server was fully started

I feel like we might be focused on optimizing the wrong thing here. This code in LanguageServerTextEditor already waits for the language server to fully start before it enables features that communicate with the language server. However, even if that code runs immediately, I would be very surprised if the highlighted section was the culprit. Adding a few event listeners and setting a couple of properties that don't do anything until the user starts interacting with the editor would not account for a massive 40 seconds of startup time. You'd need to open hundreds (maybe thousands) of editors before that particular code has an impact. It's gotta be something else. It's probably still somehow related to the language servers, but I feel like it'll be something that can be optimized in a less drastic way.

I'll play around with it a bit and report back.

**Comment by joshtynjala** [comment]
After some testing in Scout, and I made a couple of optimizations to the `RecentlyOpenedPlugin` that have a very noticeable impact on startup performance.

1) In commit be4c0ffe2ba04198e803b78b04eab2e82e42b5e2, instead of saving `ProjectReferenceVO` objects directly to the `SharedObject`, I save only the relevant fields that we actually read from the `SharedObject` when re-creating the `ProjectReferenceVO`. What I discovered in Scout was that it took a long time to read the file icon data from the `FileLocation` objects referenced by `ProjectReferenceVO`, but we don't actually need to read the icons at all. On my machine, this saved about 2.5 seconds in startup CPU time.

2) In commit 5e69ad49c4986996ec7b5f534fa27e5c0a360dfb (which has a bad description because I accidentally merged two commits and didn't notice before pushing), I modified `RecentlyOpenedPlugin` to wait for all files/projects to be opened before dispatching the `RECENT_FILES_LIST_UPDATED ` and `RECENT_PROJECT_LIST_UPDATED` events. Previously, these events were dispatched multiple times, and it became pretty expensive when opening a lot of projects/files simultaneously on startup. Now, they're dispatched only once, and it seems to save around 1 second in startup CPU time.

Note that my changes are on the master branch, and do not affect the branch that @rat-moonshine has been working on for this issue.

**Comment by rat-moonshine** [comment]
> In commit https://github.com/Moonshine-IDE/Moonshine-IDE/commit/5e69ad49c4986996ec7b5f534fa27e5c0a360dfb (which has a bad description because I accidentally merged two commits and didn't notice before pushing), I modified RecentlyOpenedPlugin to wait for all files/projects to be opened before dispatching the RECENT_FILES_LIST_UPDATED and RECENT_PROJECT_LIST_UPDATED events. Previously, these events were dispatched multiple times, and it became pretty expensive when opening a lot of projects/files simultaneously on startup. Now, they're dispatched only once, and it seems to save around 1 second in startup CPU time.

I agree to @joshtynjala and I identified same problem during my investigation to #1002 on branch. However, I only tried to tune the timing of `RECENT_PROJECT_LIST_UPDATED` event and have it dispatched only when all the projects are opened into sidebar. 

This event actually time consuming for a few seconds because it updates to two areas:
- `File > Recent Opened Project` list
- Home screen projects list

I think I can take further change suggestion from the commit https://github.com/Moonshine-IDE/Moonshine-IDE/commit/5e69ad49c4986996ec7b5f534fa27e5c0a360dfb.

**Comment by rat-moonshine** [comment]
At this point, I'm not very fond about the changes I was working on orderly execution in branch since the `master` build shown comparatively fast startup also in my tests. 

I believe it's the updates https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1002#issuecomment-1103233662 from @joshtynjala which improve the UI freeze scenario way better now. My implementation on orderly execution improvement isn't much difference than `master` now, even the `master` is a few seconds faster if the consider the startup seconds' freezing is Okay.

In branch implementation, however, the startup freeze do not demonstrate and executes stuff in an interval. But this also added its own structural complexities if we see that doesn't require since `master` improvement. 

But anyway, this may requires different system tests to get to a decision if we wants to continue with `master` fixes or go with design changes from branch.

For the time being I did not get into any structural changes of LS components (https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1002#issuecomment-1102985635). 

I also included a "Waiting for language server.." message in editor view until `Event.INIT` receives.

**Comment by JoelProminic** [comment]
I did some testing on @joshtynjala's master updates, and this seems to be a nice improvement.  Testing the Moonshine workspace, I see the sidebar populated within 3 seconds (compared to 12 in the [original tests](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1002#issuecomment-1082907406)).

I see that the language servers take a little longer to fully start, but they are loading asynchronously.  I confirmed that the editor tabs opened on startup get language server support once the language server finishes loading.

I'll ask @JustinProminic to test the updates, and I'll continue using this for my local development for now.  @rat-moonshine and I can discuss what we want to do with the updates in the 1002 branch.
